src/java/org/apache/cassandra/cql3/Vectors.java

```diff
@@ -0,0 +1,176 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.cassandra.cql3.functions.Function;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.marshal.VectorType;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+public class Vectors
+{
+    private Vectors() {}
+
+    private static AbstractType<?> elementsType(AbstractType<?> type)
+    {
+        return ((VectorType<?>) type.unwrap()).getElementsType();
+    }
+
+    private static ColumnSpecification valueSpecOf(ColumnSpecification column)
+    {
+        return new ColumnSpecification(column.ksName, column.cfName, new ColumnIdentifier("value(" + column.name + ")", true), elementsType(column.type));
+    }
+
+    public static class Literal extends Term.Raw
+    {
+        private final List<Term.Raw> elements;
+
+        public Literal(List<Term.Raw> elements)
+        {
+            this.elements = elements;
+        }
+
+        @Override
+        public TestResult testAssignment(String keyspace, ColumnSpecification receiver)
+        {
+            if (!(receiver.type instanceof VectorType))
+                return AssignmentTestable.TestResult.NOT_ASSIGNABLE;
+            VectorType<?> type = (VectorType<?>) receiver.type;
+            if (elements.size() != type.dimension)
+                return AssignmentTestable.TestResult.NOT_ASSIGNABLE;
+            ColumnSpecification valueSpec = valueSpecOf(receiver);
+            return AssignmentTestable.TestResult.testAll(receiver.ksName, valueSpec, elements);
+        }
+
+        @Override
+        public Term prepare(String keyspace, ColumnSpecification receiver) throws InvalidRequestException
+        {
+            if (!(receiver.type instanceof VectorType))
+                throw new InvalidRequestException(String.format("Invalid vector literal for %s of type %s", receiver.name, receiver.type.asCQL3Type()));
+            VectorType<?> type = (VectorType<?>) receiver.type;
+            if (elements.size() != type.dimension)
+                throw new InvalidRequestException(String.format("Invalid vector literal for %s of type %s; expected %d elements, but given %d", receiver.name, receiver.type.asCQL3Type(), type.dimension, elements.size()));
```

Comment:
Maybe `expected %d dimensions` instead of `expected %s elements`?

Reply:
in this context I feel `elements` is more correct.  I could say "expected a vector of dimension %d, but given %d" if you think thats best

Reply:
Yeah, since we are talking about a vector literal, and not a vector, elements might make more sense.